### PR TITLE
Example usage of the ask-tell interface in syne-tune

### PR DESCRIPTION
**Pull request by jgolebiowski**
*Description of changes:*
What: Setup an ask-tell interface like in scikit-optimize ([library](https://github.com/scikit-optimize/scikit-optimize), [issue that asks for it](https://github.com/scikit-optimize/scikit-optimize/issues/68) in scikit-opt) to enable users with very specific requirements to plug into the HPO engines implemented in syne-tune. This work would include providing an example that shows how to set up this mode of operation by calling a trial scheduler suggest and communicating results with on_trial_results.

Why: In order to use syne-tune at the moment, user needs to wrap their function evaluator into a script that can be executed by syne-tune. In some cases, experiments needed for function evaluations can be more complex and require extra orchestration (example vary from setting up jobs on non-aws clusters to runnig physical lab experiments)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/launch_ask_tell_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+def save_restart_with_gp() -> TrialResult:
+    metric, mode, config_space, max_iterations = get_objective()
+    scheduler = AskTellScheduler(base_scheduler=BayesianOptimization(config_space, metric=metric, mode=mode))
+    for iter in range(int(max_iterations / 2)):
+        trial_suggestion = scheduler.ask()
+        test_result = target_function(**trial_suggestion.config)
+        scheduler.tell(trial_suggestion, {metric: test_result})
+
+    scheduler.save("scheduler-checkpoint.dill")
```
Comment (geoalgo) [inline examples/launch_ask_tell_scheduler.py:159]: If we just serialize with dill, couldnt we skip the save/load method and directly dill the scheduler to simplify the example?

File: examples/launch_ask_tell_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    metric = "mean_loss"
+    mode = "min"
+    max_iterations = 100
+    config_space = {
+        "x": uniform(-1, 1),
+    }
+    return metric, mode, config_space, max_iterations
+
+
+def inspect_objective():
```
Comment (geoalgo) [inline examples/launch_ask_tell_scheduler.py:120]: plot_objective might me a better name.

File: examples/launch_ask_tell_schedululer_hyperband.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (geoalgo) [inline examples/launch_ask_tell_schedululer_hyperband.py:1]: There is a funny typo in the name of the file :-)

File: examples/launch_ask_tell_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+import datetime
```
Comment (geoalgo) [inline examples/launch_ask_tell_scheduler.py:13]: It could be helpful to add a comment at the top of the file to briefly explains what this example does.

**Review by geoalgo — COMMENTED** [review]
LGTM, thanks for providing this the code is very clean I just have minor comments.

**Comment by mseeger** [comment]
Please also update `examples.rst`

File: examples/launch_ask_tell_scheduler.py (hunk 4 above)
Comment (jgolebiowski) [reply examples/launch_ask_tell_scheduler.py:24]: added

File: examples/launch_ask_tell_scheduler.py (hunk 2 above)
Comment (jgolebiowski) [reply examples/launch_ask_tell_scheduler.py]: makes sense, chnaged

File: examples/launch_ask_tell_scheduler.py (hunk 1 above)
Comment (jgolebiowski) [reply examples/launch_ask_tell_scheduler.py]: Good point, I thought making it a function is easier but I buy the shorter=simpler argument so moved it to the function.

**Review by geoalgo — APPROVED** [review]

**Pull request merged**